service/frontend/versionChecker.go

```diff
@@ -0,0 +1,148 @@
+package frontend
+
+import (
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/api/persistenceblobs/v1"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"runtime"
+	"time"
+
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/version/check"
+)
+
+const VersionCheckInterval = 24 * time.Hour
+
+type VersionChecker struct {
+	resource.Resource
+	params       *resource.BootstrapParams
+	config       *Config
+	shutdownChan chan struct{}
+	metricsScope metrics.Scope
+}
+
+func NewVersionChecker(
+	resource resource.Resource,
+	params *resource.BootstrapParams,
+	config *Config,
+) *VersionChecker {
+	return &VersionChecker{Resource: resource, params: params, config: config, shutdownChan: make(chan struct{}),
+		metricsScope: resource.GetMetricsClient().Scope(metrics.VersionCheckScope)}
+}
+
+func (vc *VersionChecker) Start() {
+	if vc.config.EnableServerVersionCheck() {
+		go vc.versionCheckLoop()
+	}
+}
+
+func (vc *VersionChecker) Stop() {
+	if vc.config.EnableServerVersionCheck() {
+		close(vc.shutdownChan)
+	}
+}
```

Comment:
what if the start / stop function are called multiple times?

usually the start will only be called by initialization logic
there is no guarantee for stop

Reply:
As agreed up on offline, we'll ignore this for now and will have a broader discussion later.

Reply:
seems that all `start` / `stop` functions are now *not* checking at most once execution.

let me schedule a meeting to find out why and possibly add the at most once check back

Reply:
just FYI, if the stop is called >= 2 times, it will panic?

Reply:
how about adding a https://golang.org/pkg/sync/#Once.Do for now